job-runner/chain/traverser.go

```diff
@@ -320,186 +358,216 @@ func (t *traverser) Status() (proto.JobChainStatus, error) {
+	jLogger := t.logger.WithFields(log.Fields{"job_id": job.Id})
 	jl := proto.JobLog{
-		RequestId:  t.chain.RequestId(),
-		JobId:      pJob.Id,
-		Name:       pJob.Name,
-		Type:       pJob.Type,
-		SequenceId: pJob.SequenceId,
-		StartedAt:  0, // zero because the job never ran
-		FinishedAt: 0,
-		State:      pJob.State,
-		Exit:       1,
-		Error:      err.Error(),
+		RequestId:   t.chain.RequestId(),
+		JobId:       job.Id,
+		Name:        job.Name,
+		Type:        job.Type,
+		Try:         t.chain.TotalTries(job.Id),
+		SequenceId:  job.SequenceId,
+		SequenceTry: t.chain.SequenceTries(job.Id),
+		StartedAt:   0, // zero because the job never ran
+		FinishedAt:  0,
+		State:       job.State,
+		Exit:        1,
+		Error:       err.Error(),
 	}
 
 	if err != nil {
 		jl.Error = err.Error()
 	}
 
 	// Send the JL to the RM.
-	err = t.rmc.CreateJL(t.chain.RequestId(), jl)
+	err = retry.Do(t.jlTries, t.jlRetryWait,
+		func() error {
+			return t.rmc.CreateJL(t.chain.RequestId(), jl)
+		},
+		nil,
+	)
 	if err != nil {
-		jLogger.Errorf("problem sending job log (%#v) to the RM: %s", jl, err)
+		jLogger.Errorf("problem sending job log (%#v) to the Request Manager: %s", jl, err)
+		return
 	}
 }
 
-// retrySequence retries a sequence by doing the following
-//   1. Mark the failed job and all previously completed jobs in sequence to PENDING.
-//   2. Decrement the number of jobs that were previously ran (failed job +
-//   previously completed jobs in sequence) from the chain job count.
-//   3. Increment retry count for the sequence
-//   4. Enqueue the first job in the sequence to be ran again. It'll
-//   subsequently enqueue the remaining jobs in the sequence.
-func (t *traverser) retrySequence(failedJob proto.Job) error {
-	sequenceStartJob := t.chain.SequenceStartJob(failedJob)
-
-	// sequenceJobsToRetry is a list containing the failed job and all previously
-	// completed jobs in the sequence. For example, if job C of A -> B -> C -> D
-	// fails, then A and B are the previously completed jobs and C is the failed
-	// job. So, jobs A, B, and C will be added to sequenceJobsToRetry. D will not be
-	// added because it was never ran.
-	var sequenceJobsToRetry []proto.Job
-	sequenceJobsToRetry = append(sequenceJobsToRetry, failedJob)
-	sequenceJobsCompleted := t.sequenceJobsCompleted(sequenceStartJob)
-	sequenceJobsToRetry = append(sequenceJobsCompleted, sequenceJobsCompleted...)
-
-	// Mark all jobs to retry to PENDING state
-	for _, job := range sequenceJobsToRetry {
-		t.chain.SetJobState(job.Id, proto.STATE_PENDING)
+// shutdown suspends the running chain by stopping all its jobs and sending
+// a suspended job chain to the Request Manager.
+func (t *traverser) shutdown() {
+	// Don't do anything if the traverser has already finished running, or has
+	// already been stopped or suspended.
+	t.stopMux.Lock()
+	select {
+	case <-t.doneChan:
+		t.stopMux.Unlock()
+		return
+	default:
 	}
+	if t.stopped || t.suspended {
```

Comment:
I think we only need to lock and check these bool vars, don't need to check chan.